### DISCUSSION TITLE
Conversion code is ready for new cache implementation (2.0 soon-ish?)

### DISCUSSION THREAD

**Discussion by ABaumher** [ABaumher/galaxy-integration-steam#41] [Announcements]
I moved my code off of start the conversion to `Conversion_ready` to signify the change in status. The code is not finished, but it's ready for implementing the last few calls and cleaning up the cache. This is up at https://github.com/ABaumher/galaxy-integration-steam/tree/conversion_ready

A relatively brief rundown of the code: @urwrstkn8mare 's fixes are applied to how we do the OS-dependent checks for launching games, etc. This streamlines the system-specific stuff in the Plugin. The login code is now cleaned up as well - all UI-based logic is handled in the "view" class and the plugin interacts with that and the "model" to go back and forth and get all that information. The Plugin is our "controller" because with the OS fixes outlined above, it was essentially just calling the controller and that was redundant. All of the remaining calls just go straight to the model and pull that. The model contains a "local persistent cache" which functions as a collection of all the caches we use and provides one nice way to interface with them, and a "parser" which converts input data to messages, sends them off to steam's servers, and returns the result message, in a nice, clean call. Receiving messages is also simpler: there are two loops running for messages: one that simply receives them and hands them off to their caller, and another which handles "unsolicited messages" and stores any useful information in the "local persistent cache". Any call to the model will perform some combination of looking up the data in the cache and/or requesting it from steam's servers. 

A simple example of unsolicited messages are the messages sent alongside ClientLoginResponse. Steam sends us a Multi (aka collection of messages) and only one (the LoginResponse) is expected and sent back to the model. The rest are sent to the cache to handle. One of these is in particular is really important: the LicenseList. When GOG asks us for the games the user owns, that data is obtained indirectly from that license list (it's used in a few more steam calls). The cache needs to handle these messages.

files preceeded with `DEPRECATED_` are there to help port the existing code over, but should be removed (i just didn't want to have to keep getting these from git history so i left them in). They are extremely broken as a result. 

As for the state of the code: 
**Implemented**
1. All login calls are complete. 
2. All local OS-dependent code is implemented 
3. Achievements will now fire off per-game and do not have a nonsensical job list
4. User tags and related are obtained.
5. Receive loop is implemented
6. Unsolicited messages passed to cache for processing
7. Prototyped out a means to use a login without providing the password directly. The user can take a public key steam provides and encipher their password on their own using whatever tool they trust. 
<hr/>

**Not implemented**

1. Bulletproofing: 
  a) When the client logs off, the code is in place to handle it but conflicting methods of doing so.
  b) I've probably borked something somewhere. Will need rigorous testing and probably unit tests.
2. Caches - I just removed all of them because they were varying states of broken
3. Processing unsolicited messages - as a result of the lack of caches
4. Remaining calls in the model - as a result of lack of cached unsoliticed message results (and not knowing what the hell half of the messages we receive are caused by)
5. On Completed calls for Achievements and tags. They exist in the model but are no-ops (aka `pass`) These could be used to push changes to the cache or update the persistent cache gog has. 
6. The HTML changes for the users that won't input their password into GOG's integrated browser. The code is in place in python but the HTML/JS doesn't have the display for it. 
7. Extra error handling and message handling in the HTML. Error messages are now passed to the HTML as url parameters but aren't currently handled. 

Which leads me to where we are: (re)implementing the caches. @don-de-marco you now have free reign to reimplement all the caches in a way that isn't absolutely terrible. bring as much as you can from your fixes, obviously, but now you can actually implement it without the old code stabbing you in the back at every opportunity. We can use this opportunity to log more info about mystery calls to address #38. I'm aware this is a ridiculous ask, but at this point you are the expert on the caches so i'll let you take the lead on them. I can help with whatever you need me to do to help you get them cleaned up and implementing so they are as quick/non-blocking as possible.

@urwrstkn8mare @UncleGoogle @wwMRd @pippo-san @slo please take this opportunity to review the code changes. It may be easier to just look at the code as something new than compare it to the old code. @engiefox, since you're probably out there just waiting for something to break, you can take a look as well. I would not recommend trying to run it (it'll throw a NotImplemented pretty quickly or just straight up break), but assuming you can python, you could try breaking it in python's interactive environment.

**Comment by ABaumher** [comment]
Also, worth pointing out some json.loads could probably be replaced with vdf.loads so our code is cleaner but we'll need to log the raw strings so we can see if that's the case. 

Also i need to document the code better. 

Final also: the names in the protobuf socket handler are essentially the RPC calls for those functions,  which is why they're 'TitleCase' and not 'snake_case'. They can be renamed later but it made it easier when i was implementing to know what request/ response i should be using. 

I'm holding off on working on this for a little bit so everyone can catch up or ask questions or point out whatever dumb mistakes I've made. Don't hesitate to lmk if something is confusing or poorly written or straight up wrong. 

I'm not 100% on how python does variable capturing so please double check any local functions or lambdas i use. Afaik none of the captures are in loops so that gotcha is not an issue but I've learned that python often does things i don't expect and this seems like it could be one of those things i run into

**Comment by urwrstkn8mare** [comment]
is `conversion_updates` branch just for changes to `conversion_ready`? why not have us review `conversion_updates`?

**Reply by ABaumher** [reply]
I was trying to freeze the ready code but then i got impatient started updating it. That was supposed to stay local lol. But if you want to review that, that's fine as well, it's just more akin to a "nightly" build
